File: gasp/layer_greedy.py
"""Layer-building constructive front-stage (VNS-inspired).

EXPERIMENTAL STATUS (weak / reuse version): falsified as an improvement,
kept behind a flag (GASPParams.layer_greedy, default off) at zero cost.
On heterogeneous BR instances the layer construction, even with mixed
secondary types filling the border gaps, produces a worse initial
solution than the best of the eight existing sorting rules already tried
by GASP's initial_solution (e.g. on thpack7-001: layer 69% vs best
classical sort 79% of base filling). An earlier positive reading (+0.65
points) came from comparing the layer against a SINGLE density sort
rather than against GASP's full sorting portfolio. The genuine VNS gain
requires the STRONG version -- exact 2D-knapsack per layer, optimal
slab thickness, stacking with backtracking and vertical stability --
which is a different algorithm, not a front-stage reusing the EP greedy.
The weak version is retained as a documented negative result and a
scaffold for the strong one.

Original design rationale follows.

Container Loading on heterogeneous instances (many box types) is where
the score-ordering paradigm loses ground to layer-based methods such as
the VNS of Parreno et al.: those split a 20-type problem into stacked
*layers*, each a near-homogeneous 2D sub-problem that tiles almost
perfectly, and stacking full layers is trivial. This module adds that
idea as a FRONT-STAGE before the existing EP greedy, without replacing
it.

Design (weak / reuse version)
-----------------------------
1. Type guard. Count distinct box types; below a threshold the layer
   idea brings nothing (few-type instances are already handled well and
   the bottleneck there is selection, not packing), so we delegate
   entirely to the classical EP greedy. This is the "se non ci sono i
   tipi usa la vecchia" rule.
2. Layer type selection. Among the available types, pick the one whose
   layer (a slab of height equal to one of its dimensions) tiles the
   base best, scored by axial fit times profit density.
3. Layer fill. Fill a height-h slab by calling the EP greedy RESTRICTED
   to that type (optionally a second type that fits in the same height),
   then translate the placements up by the current stack height.
4. Stack and repeat over the residual height; a final pass of the
   classical EP greedy fills the gaps with the remaining items.

Everything reuses the existing geometry, EP machinery and merit
functions; the layer stage only decides WHICH items, in WHICH thin
sub-knapsack, in WHICH order.
"""

from __future__ import annotations

from collections import defaultdict
from typing import Dict, List, Optional, Tuple

from .geometry import Item, Knapsack, Packing, Placement
from .greedy import ep_kph


def _type_key(it: Item) -> Tuple[int, int, int]:
    return (it.w, it.d, it.h)


def _heights(it: Item, ks: Knapsack, allow_rotation: bool) -> List[int]:
    """Distinct feasible heights this item can present (its stacking
    dimension), one per usable rotation."""
    hs = {r[2] for r in it.rotations(allow_rotation, ks.is_3d)
          if r[0] <= ks.W and r[1] <= ks.D and r[2] <= ks.H}
    return sorted(hs)


def _axial_fit_layer(it: Item, h: int, ks: Knapsack,
                     allow_rotation: bool) -> int:
    """How many copies of `it` tile the W x D base in a slab of height
    h, over rotations that present exactly height h."""
    best = 0
    for (w, d, rh) in it.rotations(allow_rotation, ks.is_3d):
        if rh != h or w > ks.W or d > ks.D:
            continue
        best = max(best, (ks.W // w) * (ks.D // d))
    return best


class LayerGreedy:
    """Layer front-stage with a type guard, delegating to ep_kph."""

    def __init__(self, knapsack: Knapsack, criterion: str = "RS",
                 allow_rotation: bool = True,
                 type_threshold: int = 6,
                 min_layer_fill: float = 0.55,
                 max_second_types: int = 1):
        self.ks = knapsack
        self.criterion = criterion
        self.allow_rotation = allow_rotation
        self.type_threshold = type_threshold
        self.min_layer_fill = min_layer_fill
        self.max_second_types = max_second_types

    # ------------------------------------------------------------------
    def run(self, items: List[Item]) -> Packing:
        types = defaultdict(list)
        for it in items:
            types[_type_key(it)].append(it)

        # ---- type guard: few types -> classical EP greedy
        if not self.ks.is_3d or len(types) <= self.type_threshold:
            return ep_kph(self._sorted(items), self.ks, self.criterion,
                          self.allow_rotation)

        # ---- layer construction along the H axis
        placements: List[Placement] = []
        avail = {k: list(v) for k, v in types.items()}
        z_base = 0
        H = self.ks.H

        while z_base < H:
            rem_h = H - z_base
            choice = self._pick_layer_type(avail, rem_h)
            if choice is None:
                break
            key, h = choice
            slab = Knapsack(self.ks.W, self.ks.D, h)

            # items for this layer: the chosen type, plus optionally a
            # second compatible type sharing the same height
            layer_items = list(avail[key])
            second = self._second_type(avail, key, h)
            for k2 in second:
                layer_items += avail[k2]

            packing = ep_kph(self._sorted(layer_items), slab,
                             self.criterion, self.allow_rotation)
            fill = packing.used_volume / slab.volume if slab.volume else 0.0
            if not packing.placements or fill < self.min_layer_fill:
                # layer too sparse: stop the layer stage, let the final
                # EP pass handle the rest in the residual box
                break

            used_ids = packing.loaded_ids
            for p in packing.placements:
                placements.append(Placement(p.item, p.x, p.y, p.z + z_base,
                                            p.w, p.d, p.h))
            # consume used items
            for k in [key, *second]:
                avail[k] = [it for it in avail[k] if it.idx not in used_ids]
            avail = {k: v for k, v in avail.items() if v}
            z_base += h
            if not avail:
                break

        # ---- final EP pass: fill the residual top box with whatever
        # items remain, on top of the stacked layers
        remaining = [it for v in avail.values() for it in v]
        if remaining and z_base < H:
            placements = self._fill_residual(placements, remaining, z_base)

        return Packing(self.ks, placements)

    # ------------------------------------------------------------------
    def _sorted(self, items: List[Item]) -> List[Item]:
        # density-first, a sane default order for the EP greedy
        meas = (lambda i: i.volume) if self.ks.is_3d else (lambda i: i.base_area)
        return sorted(items, key=lambda i: -i.profit / meas(i))

    def _pick_layer_type(self, avail: Dict, rem_h: int):
        """Choose (type_key, height) maximising how well the layer tiles
        the base, among types that fit in the remaining height. We
        reward base COVERAGE (fraction of W x D the placed copies cover)
        times profit density, and prefer thinner layers at equal
        coverage: a thin, well-tiled slab is what makes stacking pay."""
        base = self.ks.W * self.ks.D
        best, best_score = None, 0.0
        for key, members in avail.items():
            if not members:
                continue
            rep = members[0]
            for (w, d, h) in rep.rotations(self.allow_rotation, self.ks.is_3d):
                if h > rem_h or w > self.ks.W or d > self.ks.D:
                    continue
                fit = (self.ks.W // w) * (self.ks.D // d)
                if fit == 0:
                    continue
                copies = min(len(members), fit)
                coverage = copies * w * d / base          # in [0, 1]
                if coverage < 0.35:
                    continue                              # too sparse a layer
                dens = rep.profit / rep.volume
                # thinner layers preferred at equal coverage/density
                score = coverage * dens / h
                if score > best_score:
                    best_score, best = score, (key, h)
        return best

    def _second_type(self, avail: Dict, key, h: int) -> List:
        """Optionally pick compatible secondary type(s) presenting the
        same layer height, to fill the base gaps the primary leaves."""
        if self.max_second_types <= 0:
            return []
        out = []
        for k2, members in avail.items():
            if k2 == key or not members:
                continue
            rep = members[0]
            if h in _heights(rep, self.ks, self.allow_rotation):
                out.append((rep.profit / rep.volume, k2))
        out.sort(reverse=True)
        return [k2 for _, k2 in out[:self.max_second_types]]

    def _fill_residual(self, placements: List[Placement],
                       remaining: List[Item], z_base: int) -> List[Placement]:
        """Fill the residual top box [z_base, H) with the leftover items
        using the classical EP greedy, then lift the result."""
        top = Knapsack(self.ks.W, self.ks.D, self.ks.H - z_base)
        packing = ep_kph(self._sorted(remaining), top, self.criterion,
                         self.allow_rotation)
        for p in packing.placements:
            placements.append(Placement(p.item, p.x, p.y, p.z + z_base,
                                        p.w, p.d, p.h))
        return placements

File: gasp_java/src/main/java/gasp/greedy/LayerGreedy.java
package gasp.greedy;

import gasp.geometry.Item;
import gasp.geometry.Knapsack;
import gasp.geometry.Packing;
import gasp.geometry.Placement;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.HashSet;
import java.util.Collections;

public class LayerGreedy {

    private final Knapsack ks;
    private final MeritCriterion criterion;
    private final boolean allowRotation;
    private final int typeThreshold;
    private final double minLayerFill;
    private final int maxSecondTypes;

    public LayerGreedy(Knapsack ks, MeritCriterion criterion, boolean allowRotation, int typeThreshold, double minLayerFill, int maxSecondTypes) {
        this.ks = ks;
        this.criterion = criterion;
        this.allowRotation = allowRotation;
        this.typeThreshold = typeThreshold;
        this.minLayerFill = minLayerFill;
        this.maxSecondTypes = maxSecondTypes;
    }

    private static class TypeKey {
        int w, d, h;
        public TypeKey(int w, int d, int h) {
            this.w = w; this.d = d; this.h = h;
        }
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            TypeKey typeKey = (TypeKey) o;
            return w == typeKey.w && d == typeKey.d && h == typeKey.h;
        }
        @Override
        public int hashCode() {
            int result = w;
            result = 31 * result + d;
            result = 31 * result + h;
            return result;
        }
    }

    public Packing run(List<Item> items) {
        Map<TypeKey, List<Item>> types = new java.util.LinkedHashMap<>();
        for (Item it : items) {
            TypeKey key = new TypeKey(it.w(), it.d(), it.h());
            types.computeIfAbsent(key, k -> new ArrayList<>()).add(it);
        }

        if (!ks.is3D() || types.size() <= typeThreshold) {
            return new GreedyState(ks, criterion, allowRotation).run(sorted(items));
        }

        List<Placement> placements = new ArrayList<>();
        Map<TypeKey, List<Item>> avail = new java.util.LinkedHashMap<>(types);
        int zBase = 0;
        int H = ks.H();

        while (zBase < H) {
            int remH = H - zBase;
            LayerChoice choice = pickLayerType(avail, remH);
            if (choice == null) {
                break;
            }
            TypeKey key = choice.key;
            int h = choice.h;
            
            Knapsack slab = new Knapsack(ks.W(), ks.D(), h);
            List<Item> layerItems = new ArrayList<>(avail.get(key));
            List<TypeKey> second = secondType(avail, key, h);
            for (TypeKey k2 : second) {
                layerItems.addAll(avail.get(k2));
            }

            Packing packing = new GreedyState(slab, criterion, allowRotation).run(sorted(layerItems));
            double fill = slab.volume() > 0 ? (double) packing.usedVolume() / slab.volume() : 0.0;
            
            if (packing.getPlacements().isEmpty() || fill < minLayerFill) {
                break;
            }

            Set<Integer> usedIds = new HashSet<>();
            for (Placement p : packing.getPlacements()) {
                usedIds.add(p.item().idx());
                placements.add(new Placement(p.item(), p.x(), p.y(), p.z() + zBase, p.w(), p.d(), p.h()));
            }

            List<TypeKey> toUpdate = new ArrayList<>();
            toUpdate.add(key);
            toUpdate.addAll(second);

            for (TypeKey k : toUpdate) {
                List<Item> remain = new ArrayList<>();
                for (Item it : avail.get(k)) {
                    if (!usedIds.contains(it.idx())) {
                        remain.add(it);
                    }
                }
                if (remain.isEmpty()) {
                    avail.remove(k);
                } else {
                    avail.put(k, remain);
                }
            }

            zBase += h;
            if (avail.isEmpty()) {
                break;
            }
        }

        List<Item> remaining = new ArrayList<>();
        for (List<Item> v : avail.values()) {
            remaining.addAll(v);
        }

        if (!remaining.isEmpty() && zBase < H) {
            Knapsack top = new Knapsack(ks.W(), ks.D(), H - zBase);
            Packing packing = new GreedyState(top, criterion, allowRotation).run(sorted(remaining));
            for (Placement p : packing.getPlacements()) {
                placements.add(new Placement(p.item(), p.x(), p.y(), p.z() + zBase, p.w(), p.d(), p.h()));
            }
        }

        return new Packing(ks, placements);
    }

    private List<Item> sorted(List<Item> items) {
        List<Item> sortedItems = new ArrayList<>(items);
        sortedItems.sort((i1, i2) -> {
            double meas1 = ks.is3D() ? i1.volume() : i1.baseArea();
            double meas2 = ks.is3D() ? i2.volume() : i2.baseArea();
            double val1 = -i1.profit() / meas1;
            double val2 = -i2.profit() / meas2;
            return Double.compare(val1, val2);
        });
        return sortedItems;
    }

    private static class LayerChoice {
        TypeKey key;
        int h;
        LayerChoice(TypeKey k, int h) { this.key = k; this.h = h; }
    }

    private LayerChoice pickLayerType(Map<TypeKey, List<Item>> avail, int remH) {
        long base = (long) ks.W() * ks.D();
        LayerChoice best = null;
        double bestScore = 0.0;

        for (Map.Entry<TypeKey, List<Item>> entry : avail.entrySet()) {
            TypeKey key = entry.getKey();
            List<Item> members = entry.getValue();
            if (members.isEmpty()) continue;
            Item rep = members.get(0);

            for (int[] rot : rep.rotations(allowRotation, ks.is3D())) {
                int w = rot[0], d = rot[1], h = rot[2];
                if (h > remH || w > ks.W() || d > ks.D()) continue;

                int fit = (ks.W() / w) * (ks.D() / d);
                if (fit == 0) continue;

                int copies = Math.min(members.size(), fit);
                double coverage = (double) (copies * w * d) / base;
                if (coverage < 0.35) continue;

                double dens = (double) rep.profit() / rep.volume();
                double score = coverage * dens / h;
                if (score > bestScore) {
                    bestScore = score;
                    best = new LayerChoice(key, h);
                }
            }
        }
        return best;
    }

    private List<Integer> getHeights(Item it) {
        Set<Integer> hs = new HashSet<>();
        for (int[] rot : it.rotations(allowRotation, ks.is3D())) {
            if (rot[0] <= ks.W() && rot[1] <= ks.D() && rot[2] <= ks.H()) {
                hs.add(rot[2]);
            }
        }
        List<Integer> list = new ArrayList<>(hs);
        Collections.sort(list);
        return list;
    }

    private static class SecondTypeScore implements Comparable<SecondTypeScore> {
        double score;
        TypeKey key;
        SecondTypeScore(double s, TypeKey k) { score = s; key = k; }
        @Override
        public int compareTo(SecondTypeScore o) {
            return Double.compare(o.score, this.score); // Reverse order
        }
    }

    private List<TypeKey> secondType(Map<TypeKey, List<Item>> avail, TypeKey key, int h) {
        if (maxSecondTypes <= 0) return new ArrayList<>();
        List<SecondTypeScore> out = new ArrayList<>();
        
        for (Map.Entry<TypeKey, List<Item>> entry : avail.entrySet()) {
            TypeKey k2 = entry.getKey();
            if (k2.equals(key)) continue;
            List<Item> members = entry.getValue();
            if (members.isEmpty()) continue;
            
            Item rep = members.get(0);
            if (getHeights(rep).contains(h)) {
                double score = (double) rep.profit() / rep.volume();
                out.add(new SecondTypeScore(score, k2));
            }
        }
        Collections.sort(out);
        List<TypeKey> result = new ArrayList<>();
        for (int i = 0; i < Math.min(maxSecondTypes, out.size()); i++) {
            result.add(out.get(i).key);
        }
        return result;
    }
}
